From 3037f62352831f0cc648d24f8750804b76de4beb Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Wed, 9 Sep 2015 17:28:15 -0700
Subject: [PATCH 637/827] drm: xilinx: dp_sub: Fix the scaling factor programming

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Use the correct scaling factor for 3rd component. It should be 8bit
for YUV format as well. Also, the offset wasn't being added correctly.
Add the offset correctly, so that all values are updated.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit acf1831bcd2af0e8498b9f1afe690de30f978623)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_dp_sub.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_dp_sub.c b/drivers/gpu/drm/xilinx/xilinx_drm_dp_sub.c
index fb3f441..9eab327 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_dp_sub.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_dp_sub.c
@@ -420,7 +420,7 @@ static const struct xilinx_drm_dp_sub_fmt blend_output_fmts[] = {
 		.rgb		= false,
 		.sf[0]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 		.sf[1]		= XILINX_DP_SUB_AV_BUF_4BIT_SF,
-		.sf[2]		= XILINX_DP_SUB_AV_BUF_NULL_SF
+		.sf[2]		= XILINX_DP_SUB_AV_BUF_4BIT_SF,
 	}
 };
 
@@ -451,28 +451,28 @@ static const struct xilinx_drm_dp_sub_fmt av_buf_vid_fmts[] = {
 		.rgb		= false,
 		.sf[0]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 		.sf[1]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
-		.sf[2]		= XILINX_DP_SUB_AV_BUF_NULL_SF
+		.sf[2]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 	}, {
 		.drm_fmt	= DRM_FORMAT_YUYV,
 		.dp_sub_fmt	= XILINX_DP_SUB_AV_BUF_FORMAT_NL_VID_VYUY,
 		.rgb		= false,
 		.sf[0]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 		.sf[1]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
-		.sf[2]		= XILINX_DP_SUB_AV_BUF_NULL_SF
+		.sf[2]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 	}, {
 		.drm_fmt	= DRM_FORMAT_UYVY,
 		.dp_sub_fmt	= XILINX_DP_SUB_AV_BUF_FORMAT_NL_VID_YVYU,
 		.rgb		= false,
 		.sf[0]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 		.sf[1]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
-		.sf[2]		= XILINX_DP_SUB_AV_BUF_NULL_SF
+		.sf[2]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 	}, {
 		.drm_fmt	= DRM_FORMAT_VYUY,
 		.dp_sub_fmt	= XILINX_DP_SUB_AV_BUF_FORMAT_NL_VID_YUYV,
 		.rgb		= false,
 		.sf[0]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 		.sf[1]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
-		.sf[2]		= XILINX_DP_SUB_AV_BUF_NULL_SF
+		.sf[2]		= XILINX_DP_SUB_AV_BUF_8BIT_SF,
 	}, {
 		.drm_fmt	= DRM_FORMAT_BGR888,
 		.dp_sub_fmt	= XILINX_DP_SUB_AV_BUF_FORMAT_NL_VID_RGB888,
@@ -808,11 +808,11 @@ xilinx_drm_dp_sub_av_buf_init_sf(struct xilinx_drm_dp_sub_av_buf *av_buf,
 
 	offset = XILINX_DP_SUB_AV_BUF_GFX_COMP0_SF;
 	for (i = 0; i < XILINX_DP_SUB_AV_BUF_NUM_SF; i++)
-		xilinx_drm_writel(av_buf->base, offset, gfx_fmt->sf[i]);
+		xilinx_drm_writel(av_buf->base, offset + i * 4, gfx_fmt->sf[i]);
 
 	offset = XILINX_DP_SUB_AV_BUF_VID_COMP0_SF;
 	for (i = 0; i < XILINX_DP_SUB_AV_BUF_NUM_SF; i++)
-		xilinx_drm_writel(av_buf->base, offset, vid_fmt->sf[i]);
+		xilinx_drm_writel(av_buf->base, offset + i * 4, vid_fmt->sf[i]);
 }
 
 /* Audio functions */
-- 
2.9.3

